File: android/app/src/main/kotlin/app/zen/chromium/Landing.kt
package app.zen.chromium

import android.content.Intent

/**
 * The state a Home-screen widget or a launcher shortcut asks the browser window to land in
 * (WID-07): the omnibox focused with the keyboard up ([SEARCH]), the voice search sheet listening
 * ([VOICE]), a new private tab ([PRIVATE]) or the QR scanner ([SCAN]). It travels as one string
 * extra on the intent that starts or re-enters `MainActivity`, and nothing else: the activity's
 * intent switch reads it ([of]) and hands the word to `ChromeWebView.land`. On a cold start the
 * chrome's document is still loading, so the word waits in a [Stash] and travels IN the core's
 * boot answer (`BootInfo.landing`), which `bootAndroid` applies right after `browser.start()`, in
 * the boot's own run – so the chrome's first frame already shows the requested state and the
 * restored tab is never seen first. Warm (`onNewIntent`) it goes to `window.__zenHost.land` at
 * once. No new work on the boot path: an extra, its reader, and one null read of the answer's
 * field in a start that carries none.
 *
 * The widget's `PendingIntent`s ([SearchWidgetProvider]) aim straight at `MainActivity` – a widget's
 * intent is fired as built, unlike a manifest shortcut's, which the system stamps with
 * `FLAG_ACTIVITY_CLEAR_TASK` (`ShortcutParser`) and which therefore goes through the
 * `LauncherIconActivity` trampoline with the extra carried over ([forwarded]). Pure where it can
 * be, so the parse has a JVM test.
 */
object Landing {
    /** The intent extra's name; its value one of [STATES]. */
    const val EXTRA = "app.zen.chromium.extra.LANDING"

    /** The omnibox focused for the active tab, the keyboard up (the widget's bar, the Search shortcut). */
    const val SEARCH = "search"

    /** The voice search sheet open and listening (the widget's mic). */
    const val VOICE = "voice"

    /** A new private tab in the current space (the widget's mask, the New private tab shortcut). */
    const val PRIVATE = "private"

    /** The QR scanner open (the Scan QR code shortcut). */
    const val SCAN = "scan"

    /** A plain new tab in the current space, active (the New tab shortcut). */
    const val NEW_TAB = "newTab"

    val STATES: Set<String> = setOf(SEARCH, VOICE, PRIVATE, SCAN, NEW_TAB)

    /**
     * The state a raw extra value names, or null for anything that is not one of [STATES]: an
     * absent extra, an empty string, or a word this build does not know (an older widget's intent
     * after an update, a hand-built `am start`). Case and surrounding blanks are forgiven, the
     * value the chrome receives is the canonical word.
     */
    fun parse(value: String?): String? {
        val word = value?.trim() ?: return null
        return STATES.firstOrNull { it.equals(word, ignoreCase = true) }
    }

    /**
     * The landing [intent] asks for – the extra's word, or the one its action implies
     * ([forwarded]: a third party's explicit `NEW_PRIVATE_TAB` intent aimed at `MainActivity`
     * itself is the private landing by another name, on the same path) – or null.
     */
    fun of(intent: Intent?): String? = forwarded(intent?.action, intent?.getStringExtra(EXTRA))

    /**
     * The landing an intent's word and action come to: the landing it named, or the one its
     * action implies (`PrivateBrowsing.ACTION_NEW_TAB` is the private landing by another name –
     * the shortcut that predates the extra keeps its action, and so does a pinned copy of it aimed
     * straight at `MainActivity`). What the trampoline carries over to `MainActivity` for a
     * shortcut's intent, and what `MainActivity` reads of its own. Null for a plain launcher tap.
     */
    fun forwarded(action: String?, extra: String?): String? =
        parse(extra) ?: if (action == PrivateBrowsing.ACTION_NEW_TAB) PRIVATE else null

    /**
     * Where a landing waits for the core's boot answer (`ChromeWebView.land`, [take] from
     * `Host.dispatchSync("boot")`). Until the document's core has asked for its boot, [offer]
     * keeps the word (a later offer replaces an earlier one: the last tap is what the user
     * wants) and [take] hands it into the answer, once; from then on the document is booted,
     * [offer] declines and the caller sends the word to the host global at once – the warm path.
     * [reset] is a replaced chrome document, whose core will ask for its boot again. Offered on
     * the main thread, taken on the bridge thread, hence the lock. Pure, so it has a JVM test.
     */
    class Stash {
        private val lock = Any()
        private var pending: String? = null
        private var booted = false

        /** Keep [state] for the boot answer; false once that answer has been taken (send it now). */
        fun offer(state: String): Boolean = synchronized(lock) {
            if (booted) {
                false
            } else {
                pending = state
                true
            }
        }

        /** The boot answer's landing: what waited, or null; the document counts as booted from here on. */
        fun take(): String? = synchronized(lock) {
            booted = true
            pending.also { pending = null }
        }

        /** The chrome document is being replaced: the next boot answer carries the next landing. */
        fun reset() {
            synchronized(lock) { booted = false }
        }
    }
}

File: src/android/landing.ts
import type { Tab } from '@shared/types'
import { BLANK_URL } from '@shared/url'
import type { Browser } from '@core/browser'
import type { ZenWindow } from '@core/window'
import { browserStore } from '@renderer/lib/browserStore'
import { startQrScan } from '@renderer/lib/qrScan'
import { openNewTabPageUrlbar, pushToast } from '@renderer/lib/ui'
import { startVoiceSearch } from '@renderer/lib/voiceSearch'
import { openShortcutPrivateTab, PRIVATE_TABS_UNAVAILABLE } from './privateShortcut'

/**
 * The states the search widget and the launcher's shortcuts ask the app to open in (WID-07). The
 * words are `Landing.kt`'s: one intent extra (`app.zen.chromium.extra.LANDING`), read by
 * `MainActivity.handleIntent` and handed to `ChromeWebView.land`. Cold, the chrome stashes it
 * and the boot answer carries it (`BootInfo.landing`), applied by `bootAndroid` right after
 * `browser.start()`; warm (`onNewIntent`), it reaches `window.__zenHost.land` at once.
 */
export const LANDING_STATES = ['search', 'voice', 'private', 'scan', 'newTab'] as const
export type LandingState = (typeof LANDING_STATES)[number]

/** The word as the host sent it, or null for anything that is not one of the states. */
export function parseLanding(value: unknown): LandingState | null {
  if (typeof value !== 'string') return null
  const word = value.trim()
  return LANDING_STATES.find((state) => state.toLowerCase() === word.toLowerCase()) ?? null
}

/** The surface a landing puts over its tab once the chrome is up; injected so the plan is testable. */
export interface LandingSurfaces {
  /** The omnibox open over the new tab page, bound to the tab, the keyboard following its field. */
  omnibox(tabId: string): void
  /** The voice search sheet, listening, its result loading in the tab (W2-7's path). */
  voice(tabId: string): void
  /** The QR scan sheet over the tab. */
  scan(tabId: string): void
  /** The word for a landing the core declines (the private tab on a WebView without profiles). */
  unavailable(message: string): void
}

const surfaces: LandingSurfaces = {
  // What `newtab.opened` does for a new tab on a touch layout (`useMainEvents`): new-tab mode
  // bound to the tab, attached to the bar, so what is typed navigates this tab.
  omnibox: (tabId) => openNewTabPageUrlbar(tabId, undefined, true),
  voice: (tabId) => void startVoiceSearch({ tabId }),
  scan: (tabId) => void startQrScan({ tabId }),
  // The chrome's own toast store, where the window's `toast` event ends (`useMainEvents`) – but
  // said here directly, so a word given inside the boot's run is in the first frame: the event
  // has no subscriber until after the first render and is not replayed to a late one.
  unavailable: (message) => void pushToast(message, 'error')
}

/**
 * Run `fn` once the chrome holds the core's state – at once when it does. On a cold start the
 * landing is applied by `bootAndroid` itself, from the boot answer, in the same synchronous run
 * as `browser.start()` and before the chrome has mounted (`main.tsx` renders once `bootAndroid`
 * resolves): the tab is in place for the chrome's first frame, and the surface over it is asked
 * for as soon as the chrome can answer (`startBrowserSync` fills the store in the microtasks
 * before the first render), never a frame late. Warm (`onNewIntent`) the store is filled and the
 * surface goes up in the intent's own turn.
 */
function whenChromeHasState(fn: () => void): void {
  if (browserStore.get().state !== null) {
    fn()
    return
  }
  const stop = browserStore.subscribe(() => {
    if (browserStore.get().state === null) return
    stop()
    fn()
  })
}

/**
 * Land the app in `state` (WID-07): a widget face or a launcher shortcut was tapped and the app
 * must open straight in what was asked for – the omnibox focused, the mic listening, a new
 * private tab, the QR scanner – without the previous tab ever painting.
 *
 * The rule that gives the "no flash": every landing is a NEW tab, created active in this turn
 * (`private` is the shortcut's private tab; the rest a blank tab, the phone's new tab page). Cold,
 * this runs inside `bootAndroid`, right after `browser.start()` and before React mounts, so the
 * chrome's first frame already shows the new tab and the restored one never takes a frame. Warm,
 * the next frame shows it. The surface (omnibox, voice, scan) is the chrome's own entry point,
 * called as the chrome is ready (`whenChromeHasState`).
 *
 * The tab is `fromIntent`, as the private shortcut's is (INC-01): back at its root returns to the
 * launcher and closes the tab on the way out (#117's `rootBackAction` -> `caller`) – the widget's
 * tab is one the launcher sent, and Chrome's search widget behaves the same.
 *
 * Returns the tab it opened (null: an unknown word, or the private tab declined by a WebView
 * without profiles – the shortcut's word for that, `PRIVATE_TABS_UNAVAILABLE`, is said through
 * the chrome's own surface as the chrome is ready, the way the other surfaces go up: the
 * shortcut's `browser.toast` is the window's `toast` event, and inside the boot's run that event
 * has no subscriber yet, so a cold private landing on such a WebView used to land in silence).
 */
export function landFromIntent(
  word: string,
  browser: Browser,
  win: ZenWindow,
  over: LandingSurfaces = surfaces,
  ready: (fn: () => void) => void = whenChromeHasState
): Tab | null {
  const state = parseLanding(word)
  if (state === null) {
    console.warn(`[zen] landing: not a state: ${JSON.stringify(word)}`)
    return null
  }
  if (state === 'private') {
    if (browser.state.capabilities.privateTabs) return openShortcutPrivateTab(browser, win)
    ready(() => over.unavailable(PRIVATE_TABS_UNAVAILABLE))
    return null
  }
  const tab = browser.tabs.createTab({ url: BLANK_URL, active: true, fromIntent: true }, win)
  const surface = surfaceOf(state)
  if (surface) ready(() => over[surface](tab.id))
  return tab
}

/** The surface a state puts over its tab; `newTab` and `private` have none. */
export function surfaceOf(state: LandingState): keyof LandingSurfaces | null {
  switch (state) {
    case 'search':
      return 'omnibox'
    case 'voice':
      return 'voice'
    case 'scan':
      return 'scan'
    default:
      return null
  }
}
